File: java/GreatestProductPath.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class GreatestProductPath {
    public static class Coordinate {
        int x;
        int y;
        
        public Coordinate(int x, int y) {
            this.x = x;
            this.y = y;
        }
        
        public String toString() {
            return "(" + x + ", " + y + ")";
        }
    }
    
    public static int greatestProductPath(int[][] matrix) {
        return greatestProductPath(matrix, 0, 0);
    }
    
    private static int greatestProductPath(int[][] matrix, int i, int j) {
        if (i == matrix.length - 1 && j == matrix[0].length - 1) return matrix[i][j];
        if (i >= matrix.length || j >= matrix[0].length) return 0;
        
        return matrix[i][j] * Math.max(greatestProductPath(matrix, i+1, j),
                                       greatestProductPath(matrix, i, j+1));
    }
    
    public static int greatestProductPath2(int[][] matrix) {
        List<Integer> result = new LinkedList<Integer>();
        result.add(Integer.MIN_VALUE);
        greatestProductPath2(matrix, 0, 0, 1, result);
        return result.get(0);
    }
    
    private static void greatestProductPath2(int[][] matrix, int i, int j, int product, List<Integer> result) {
        if (i >= matrix.length || j >= matrix[0].length) return;
        
        product *= matrix[i][j];
        
        if (i == matrix.length - 1 && j == matrix[0].length - 1) {
            if (result.get(0) < product) {
                result.set(0, product);
            }
            return;
        }
        
        greatestProductPath2(matrix, i+1, j, product, result);
        greatestProductPath2(matrix, i, j+1, product, result);
    }
    
    public static int greatestProductPathBuildUp(int[][] matrix) {
        return greatestProductPathBuildUp(matrix, 0, 0)[0];
    }
    
    private static int[] greatestProductPathBuildUp(int[][] matrix, int i, int j) {
        if (i >= matrix.length || j >= matrix[0].length) return null;
        
        int[] right = greatestProductPathBuildUp(matrix, i+1, j);
        int[] down = greatestProductPathBuildUp(matrix, i, j+1);
        
        if (right == null && down == null) return new int[]{matrix[i][j], matrix[i][j]};
        
        int max = Integer.MIN_VALUE;
        int min = Integer.MAX_VALUE;
        
        if (right != null) {
            max = Math.max(max, Math.max(right[0] * matrix[i][j],
                                         right[1] * matrix[i][j]));
            min = Math.min(min, Math.min(right[0] * matrix[i][j],
                                         right[1] * matrix[i][j]));
        }
        
        if (down != null) {
            max = Math.max(max, Math.max(down[0] * matrix[i][j],
                                         down[1] * matrix[i][j]));
            min = Math.min(min, Math.min(down[0] * matrix[i][j],
                                         down[1] * matrix[i][j]));
        }
        
        return new int[]{max, min};
    }
    
    public static void main(String[] args) {
        int[][] matrix = new int[][]{
            new int[]{1,2,3},
            new int[]{4,5,6},
            new int[]{7,8,-9}};
        
        System.out.println(greatestProductPathBuildUp(matrix));
    }
}

File: python3/GreatestProductPath.py
class GreatestProductPath():
    """
    Given an NxN matrix, find the path from (0,0) to (N-1, N-1) by moving either right or down that has the greatest product
    """
    def greatest_product_path(matrix):
    	return self.greatest_product_path_helper(matrix, 0, 0)

	def greatest_product_path_helper(matrix, i, j):
		if i == len(matrix)-1 and j == len(matrix[0] -1):
			return matrix[i][j]

		if i >= len(matrix) or j >= len(matrix[0]):
			return 0

		return matrix[i][j] * max(self.greatest_product_path_helper(matrix, i + 1, j), \
								  self.greatest_product_path_helper(matrix, i, j+ 1))

	def greatest_product_path2(matrix):
		result = []
		result.append(-float('inf'))
		self.greatest_product_path2_helper(matrix, 0, 0, 1, result)
		return result[0]

	def greatest_product_path2_helper(matrix, i, j, product, result):
		if i >= len(matrix) or j >= len(matrix[0]):
			return

		product *= matrix[i][j]

		if i == len(matrix) - 1 and j == len(matrix[0]) - 1:
			if result[0] < product:
				result[0] = product
			return

		self.greatest_product_path2_helper(matrix, i + 1, j, product, result)
		self.greatest_product_path2_helper(matrix, i, j + 1, product, result)


	def greatest_product_path_built_up(matrix):
		return self.greatest_product_path_built_up_helper(matrix, 0, 0)[0]

	def greatest_product_path_built_up_helper(matrix, i, j):
		if i >= len(matrix) or j >= len(matrix[0]):
			return None

		right = self.greatest_product_path_built_up_helper(matrix, i + 1, j)
		down = self.greatest_product_path_built_up_helper(matrix, i, j + 1)

		if not right and not down:
			return (matrix[i][j], matrix[i][j])

		max_val = float('inf')
		min_val = -float('inf')

		if right:
			max_val = max(max_val, max(right[0] * matrix[i][j], \
										right[1] * matrix[i][j]))
			min_val = min(min_val, min(right[0] * matrix[i][j], \
										right[1] * matrix[i][j]))

		if down:
			max_val = max(max_val, max(down[0] * matrix[i][j], \
										down[1] * matrix[i][j]))
			min_val = min(min_val, min(down[0] * matrix[i][j], \
										down[1] * matrix[i][j]))

		return (max_val, min_val)
    
